### PR TITLE
update process_question_md to move test dir files

### PR DESCRIPTION
**Pull request by amurph24**
- appends code to `process_question_md(..)` to move autograder test files. This is almost identical to `process_question_pl(..)` but doesn't copy the `test.py` file (which has the unit tests), into the student version.
TODO:
- ~~add test files to appropriate dir's~~
---
- CI tests have been updated to check for all files, and not assume a single file per problem (Cheers Bluesy1)
- `tests` dir files have been added to the expected (instructor)[https://github.com/open-resources/problem_bank_scripts/tree/file-editor_test_dir_fix/tests/test_question_templates/question_expected_outputs/instructor/q13_file-editor-input], and (student)[https://github.com/open-resources/problem_bank_scripts/tree/file-editor_test_dir_fix/tests/test_question_templates/question_expected_outputs/public/q13_file-editor-input] outputs for the file-editor template

### CODE REVIEW

**Comment by Bluesy1** [comment]
I don't have time right now, but I can do it tonight or tommorow morning - to fix the test failues we need to update the tests to not assume a single file per problem for the public and instructor tests like I did here for prarielearn output testing: https://github.com/open-resources/problem_bank_scripts/commit/9ce4c34b071d18fdb86bbc426b07a2c07ef5aa73#diff-51f37c2c9d099e0e1ce03346c3c5ebd9e10f7dcc10164db2d6365571cbbddbd8R95-R102

**Comment by amurph24** [comment]
You are more familiar with the tests for this so I'll let you handle updating the tests, I think the pathing is fixed now for the `tests/` dir

**Comment by Bluesy1** [comment]
> You are more familiar with the tests for this so I'll let you handle updating the tests, I think the pathing is fixed now for the `tests/` dir

@amurph24 Done - I noticed a bug in the test moving code while doing it, so I went ahead and fixed it too:
image

File: tests/test_problem_bank_scripts.py
Comment (firasm) [inline tests/test_problem_bank_scripts.py:210]: I'd feel more comfortable if we used the built-in `pathlib` module to create these paths if possible (rather than converting to strings and replacing. 

This is a good review: https://realpython.com/python-pathlib/, see specifically [this section](https://realpython.com/python-pathlib/#renaming-files).

File: tests/test_problem_bank_scripts.py
Comment (firasm) [inline tests/test_problem_bank_scripts.py:254]: Same comment.

**Review by firasm — CHANGES_REQUESTED** [review]
Minor comments on code - I did not check the tests myself thoroughly. might be a good idea to rebase on main and run them again before it's merged in.

File: tests/test_problem_bank_scripts.py
Comment (Bluesy1) [reply tests/test_problem_bank_scripts.py:210]: What you're linking to as far as I can tell is for *renaming* files, as per the [pathlib docs for `Path.rename(target)`](https://docs.python.org/3.11/library/pathlib.html#pathlib.Path.rename):

> Rename this file or directory to the given target, and return a new Path instance pointing to target. On Unix, if target exists and is a file, it will be replaced silently if the user has permission. On Windows, if target exists, [FileExistsError](https://docs.python.org/3.11/library/exceptions.html#FileExistsError) will be raised. target can be either a string or another path object:

  ```python
  >>> p = Path('foo')

  >>> p.open('w').write('some text')
  9
  >>> target = Path('bar')
  >>> p.rename(target)
  PosixPath('bar')
  >>> target.open().read()
  'some text'

  ```
> The target path may be absolute or relative. Relative paths are interpreted relative to the current working directory, not the directory of the Path object.
> It is implemented in terms of [os.rename()](https://docs.python.org/3.11/library/os.html#os.rename) and gives the same guarantees.
> Changed in version 3.8: Added return value, return the new Path instance.

If you would prefer to make sure we have the full path before converting it to a string, we could do the following:

```suggestion
                file, pathlib.Path(str(file.absolute()).replace(str(comparePath.absolute()), str(outputPath))), shallow=False
```

Unless you want something that (ab)uses `pathlib.Path.parent` among other tools whole bunch to attempt pull out the relative fp after the where it diverges into the expected/generated output dirs.

When I initially made this change for the prarielearn tests in #37, I tried really hard to find a better way than string casting the paths to do it, but I couldn't find any tool currently in pathlib to do it, though potentially we'll be able to do it ourselves, or though pathlib/a 3-rd party extension to pathlib whenever this happens: https://discuss.python.org/t/make-pathlib-extensible/3428

**Comment by Bluesy1** [comment]
> Minor comments on code - I did not check the tests myself thoroughly. might be a good idea to rebase on main and run them again before it's merged in.

Currently, I recommend against any sort of rebasing or merging main into this branch, its liable to cause CI to break currently, commit 95521f1f30b68b6548c66f25660695a0debc9f07 yesterday broke CI by taking the lockfile out of sync with the dependencies specified in `pyproject.toml`:
```bash
Run poetry install --with dev
Installing dependencies from lock file
Warning: poetry.lock is not consistent with pyproject.toml. You may be getting improper dependencies. Run `poetry lock [--no-update]` to fix it.

Because problem-bank-scripts depends on pandas (^2.0) which doesn't match any versions, version solving failed.
Error: Process completed with exit code 1.
```

File: tests/test_problem_bank_scripts.py
Comment (Bluesy1) [reply tests/test_problem_bank_scripts.py:210]: I suppose another solution thats effectively equivalent to using `str#replace` but is more "native pathlib" is as follows:
```python
>>> import pathlib
>>> path = pathlib.Path("tests/test_question_templates/question_expected_outputs/public/q01_multiple-choice/q01_multiple-choice.md")
>>> path
PosixPath('tests/test_question_templates/question_expected_outputs/public/q01_multiple-choice/q01_multiple-choice.md')
>>> pathlib.Path(*(part if part != "question_expected_outputs" else "question_generated_outputs" for part in path.parts))
PosixPath('tests/test_question_templates/question_generated_outputs/public/q01_multiple-choice/q01_multiple-choice.md')
```
We'd implement that here as such:

```suggestion
                file, pathlib.Path(*(part if part != "question_expected_outputs" else "question_generated_outputs" for part in file.parts)), shallow=False
```

Comment (Bluesy1) [reply tests/test_problem_bank_scripts.py:210]: Actually I just realized I missed [`pathlib.Path.relative_to(*args)`](https://docs.python.org/3.11/library/pathlib.html#pathlib.PurePath.relative_to) in the docs:
```python
>>> import pathlib
>>> outputDest = pathlib.Path("tests/test_question_templates/question_generated_outputs/")
>>> compareDest =  pathlib.Path("tests/test_question_templates/question_expected_outputs/")
>>> file = compareDest / "public/q01_multiple-choice/q01_multiple-choice.md"
>>> file
PosixPath('tests/test_question_templates/question_expected_outputs/public/q01_multiple-choice/q01_multiple-choice.md')
>>> outputDest / file.relative_to(compareDest)
PosixPath('tests/test_question_templates/question_generated_outputs/public/q01_multiple-choice/q01_multiple-choice.md')
```
which enables the following solution to the paths issue with `pathlib` as natively as it gets:
```suggestion
                file, outputDest / file.relative_to(compareDest), shallow=False
```

Comment (firasm) [reply tests/test_problem_bank_scripts.py]: Perfect! This is a great solution.

**Comment by firasm** [comment]
I think I've fixed things on main now.

I think I'll approve this and merge it in, but wait till things stabilize before cutting a new release.

**Pull request merged**
